File: leetcode/454-4sum-ii.py
# https://leetcode.com/problems/4sum-ii/ (top-interview-questions)

from typing import List

class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        count = 0
        hash_map = dict()

        for num1 in nums1:
            for num2 in nums2:
                sum_ = num1 + num2
                hash_map[sum_] = hash_map.get(sum_, 0) + 1

        for num3 in nums3:
            for num4 in nums4:
                negation = -num3-num4
                if negation in hash_map:
                    count += hash_map[negation]
        
        return count

File: leetcode/454-4sum-ii.js
// https://leetcode.com/problems/4sum-ii/ (top-interview-questions)

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number[]} nums3
 * @param {number[]} nums4
 * @return {number}
 */
var fourSumCount = function(nums1, nums2, nums3, nums4) {
  const map = {};
  for (const num1 of nums1) {
    for (const num2 of nums2) {
      const sum = num1 + num2;
      if (!map.hasOwnProperty(sum)) map[sum] = 0;
      map[sum] += 1;
    }
  }

  let count = 0;
  for (const num3 of nums3) {
    for (const num4 of nums4) {
      const negation = -(num3+num4);
      if (map.hasOwnProperty(negation)) {
        count += map[negation];
      }
    }
  }

  return count;
};
